src/sqlfluff/rules/L003.py

```diff
@@ -776,136 +830,263 @@ def _process_current_line(
+        current_line: List[BaseSegment],
+        templated_file: Optional[TemplatedFile],
+    ) -> None:
+        self.current_line = [el for el in current_line if not el.is_whitespace]
+        self.templated_file = templated_file
+
+    @property
+    def template_content(self):
+        return "".join(
+            seg.raw or getattr(seg, "source_str", "") for seg in self.current_line
+        )
+
+    def is_single_placeholder_line(self):
         count_placeholder = 0
-        for seg in current_line:
+        for seg in self.current_line:
             if seg.is_code:
                 return False
             elif seg.is_type("placeholder"):
                 count_placeholder += 1
         return count_placeholder == 1
 
-    @classmethod
-    def _is_template_block_start_line(cls, current_line, templated_file):
-        def segment_info(idx: int) -> Tuple[str, Optional[str]]:
-            """Helper function."""
-            seg = current_line[idx]
-            return seg.type, cls._get_element_template_info(seg, templated_file)
-
-        if not cls._single_placeholder_line(current_line):
-            return False
-        for idx in range(1, len(current_line)):
-            if segment_info(idx - 1) in (
-                ("placeholder", "block_start"),
-                ("placeholder", "compound"),
-                ("placeholder", "block_mid"),
-            ) and segment_info(idx) == ("indent", None):
-                return True
-        return False
+    def list_segement_and_raw_segement_types(self):
+        """Yields the tuple of seg type and underlying type were applicable."""
+        for seg in self.current_line:
+            raw_seg = self.get_raw_slices(seg)
+            yield (seg.type, raw_seg[0].slice_type if raw_seg else None)
+
+    def iterate_type_pairs(self):
+        """Produce a list of pairs of each sequenctial combo of two."""
+        iterable = self.list_segement_and_raw_segement_types()
+        a, b = itertools.tee(iterable)
+        # consume the first item in b
+        next(b, None)
+        return zip(a, b)
+
+    def is_block_start(self):
+        # Check for this sequence
+        start_blocks = (
+            ("placeholder", "block_start"),
+            ("placeholder", "compound"),
+            ("placeholder", "literal"),
+            ("placeholder", "block_mid"),
+        )
+        indent_types = (
+            ("indent", None),
+            ("newline", None),
+        )
+        valid_combos = list(
+            itertools.product(
+                start_blocks,
+                indent_types,
+            )
+        )
+        return any(pair in valid_combos for pair in self.iterate_type_pairs())
+
+    def is_block_end(self):
+        # Check for this sequence
+        dedent_types = (("dedent", None),)
+        end_block = (
+            ("placeholder", "block_end"),
+            ("placeholder", "compound"),
+            ("placeholder", "block_mid"),
+        )
+        valid_combos = list(
```

Comment:
It looks like this is going to compute the cross product every time the function is executed -- probably not a huge cost, but seems unnecessary. Think we should cache the result of this computation?

Reply:
yup can do. Will moving this to a class property have that effect?

Reply:
Yes, assignments at class level are only evaluated when that module is first imported.